crates/ruff_python_ast/src/nodes.rs

```diff
@@ -3197,37 +3254,37 @@ pub struct Parameters {
 }
 
 impl Parameters {
-    /// Returns the [`ParameterWithDefault`] with the given name, or `None` if no such [`ParameterWithDefault`] exists.
-    pub fn find(&self, name: &str) -> Option<&ParameterWithDefault> {
+    /// Returns an iterator over all non-variadic parameters included in this [`Parameters`] node.
+    ///
+    /// The variadic parameters (`.vararg` and `.kwarg`) can never have default values;
+    /// non-variadic parameters sometimes will.
+    pub fn iter_non_variadic_params(&self) -> impl Iterator<Item = &ParameterWithDefault> {
         self.posonlyargs
             .iter()
             .chain(&self.args)
             .chain(&self.kwonlyargs)
+    }
+
+    /// Returns the [`ParameterWithDefault`] with the given name, or `None` if no such [`ParameterWithDefault`] exists.
+    pub fn find(&self, name: &str) -> Option<&ParameterWithDefault> {
+        self.iter_non_variadic_params()
             .find(|arg| arg.parameter.name.as_str() == name)
     }
 
-    /// Returns `true` if a parameter with the given name included in this [`Parameters`].
-    pub fn includes(&self, name: &str) -> bool {
-        if self
-            .posonlyargs
+    /// Returns an iterator over all parameters included in this [`Parameters`] node.
+    pub fn iter_all_params(&self) -> impl Iterator<Item = AnyParameter> {
```

Comment:
I would name this `iter()` which is the first method that I normally try on any iterable type. 

You can also consider implementing 

```rust
impl<'a> IntoIterator<Item=&'a AnyParameter> for &'a Parameters {
	// 

}
```

It will probably require you to write a manual implementation of the `Iterator` or write up a very long type for the `IntoIter` type. 

What it gives you is that

```rust
for parameter in &parameters {
	/// 
}
```

works.

Reply:
> I would name this `iter()` which is the first method that I normally try on any iterable type.

I wondered about this. In the end I opted for the name I have now, as I felt like it emphasised the fact that `iter_all_params()` and `iter_non_variadic_params()` return iterators of different types. But I see your point; I'll go for the more conventional name 👍

Reply:
> or write up a very long type for the `IntoIter` type.

Seems like it would be

```rs
type ParameterIterator<'a> = Chain<
    Chain<
        Chain<
            Map<
                Chain<
                    std::slice::Iter<'a, ParameterWithDefault>,
                    std::slice::Iter<'a, ParameterWithDefault>,
                >,
                dyn FnOnce(&ParameterWithDefault) -> AnyParameterRef<'a>,
            >,
            std::option::IntoIter<AnyParameterRef<'a>>,
        >,
        Map<
            std::slice::Iter<'a, ParameterWithDefault>,
            dyn FnOnce(ParameterWithDefault) -> AnyParameterRef<'a>,
        >,
    >,
    std::option::IntoIter<AnyParameterRef<'a>>,
>;
```

Reply:
Hehe yep :D

Reply:
> The dyn part surprises me a bit

Yeah, that part is the only bit that doesn't compile, and I don't know how to make it compile :P

Reply:
It might be easier to write your own iterator type. It will require a lot of boilerplate

Reply:
I managed to get a custom iterator type working. It did indeed take a lot of boilerplate. I'm not sure that it's worth it right now, as there would only be three callsites currently where we'd be able to directly make use of the `impl IntoIterator<'a> for &'a Parameters` implementation:

<details>

```diff
diff --git a/crates/ruff_linter/src/checkers/ast/mod.rs b/crates/ruff_linter/src/checkers/ast/mod.rs
index c4794749f..566924b53 100644
--- a/crates/ruff_linter/src/checkers/ast/mod.rs
+++ b/crates/ruff_linter/src/checkers/ast/mod.rs
@@ -604,7 +604,7 @@ impl<'a> Visitor<'a> for Checker<'a> {
                     self.visit_type_params(type_params);
                 }
 
-                for parameter in parameters.iter() {
+                for parameter in &**parameters {
                     if let Some(expr) = parameter.annotation() {
                         if singledispatch && !parameter.is_variadic() {
                             self.visit_runtime_required_annotation(expr);
diff --git a/crates/ruff_python_ast/src/node.rs b/crates/ruff_python_ast/src/node.rs
index 3599e49c8..d3192e275 100644
--- a/crates/ruff_python_ast/src/node.rs
+++ b/crates/ruff_python_ast/src/node.rs
@@ -4222,7 +4222,7 @@ impl AstNode for Parameters {
     where
         V: PreorderVisitor<'a> + ?Sized,
     {
-        for parameter in self.iter() {
+        for parameter in self {
             match parameter {
                 AnyParameterRef::NonVariadic(parameter_with_default) => {
                     visitor.visit_parameter_with_default(parameter_with_default);
diff --git a/crates/ruff_python_ast/src/nodes.rs b/crates/ruff_python_ast/src/nodes.rs
index b464ffe68..568703fba 100644
--- a/crates/ruff_python_ast/src/nodes.rs
+++ b/crates/ruff_python_ast/src/nodes.rs
@@ -2,6 +2,7 @@
 
 use std::fmt;
 use std::fmt::Debug;
+use std::iter::FusedIterator;
 use std::ops::Deref;
 use std::slice::{Iter, IterMut};
 use std::sync::OnceLock;
@@ -3271,14 +3272,8 @@ impl Parameters {
     }
 
     /// Returns an iterator over all parameters included in this [`Parameters`] node.
-    pub fn iter(&self) -> impl Iterator<Item = AnyParameterRef> {
-        self.posonlyargs
-            .iter()
-            .chain(&self.args)
-            .map(AnyParameterRef::NonVariadic)
-            .chain(self.vararg.as_deref().map(AnyParameterRef::Variadic))
-            .chain(self.kwonlyargs.iter().map(AnyParameterRef::NonVariadic))
-            .chain(self.kwarg.as_deref().map(AnyParameterRef::Variadic))
+    pub fn iter(&self) -> ParametersIterator {
+        ParametersIterator::new(self)
     }
 
     /// Returns the total number of parameters included in this [`Parameters`] node.
@@ -3305,6 +3300,81 @@ impl Parameters {
     }
 }
 
+pub struct ParametersIterator<'a> {
+    parameters: &'a Parameters,
+    index: usize,
+}
+
+impl<'a> ParametersIterator<'a> {
+    fn new(parameters: &'a Parameters) -> Self {
+        Self {
+            parameters,
+            index: 0,
+        }
+    }
+}
+
+impl<'a> Iterator for ParametersIterator<'a> {
+    type Item = AnyParameterRef<'a>;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if self.index >= self.parameters.len() {
+            return None;
+        }
+
+        let mut index = self.index;
+        self.index += 1;
+
+        let Parameters {
+            range: _,
+            posonlyargs,
+            args,
+            vararg,
+            kwonlyargs,
+            kwarg,
+        } = self.parameters;
+
+        if index < posonlyargs.len() {
+            return Some(AnyParameterRef::NonVariadic(&posonlyargs[index]));
+        }
+        index -= posonlyargs.len();
+        if index < args.len() {
+            return Some(AnyParameterRef::NonVariadic(&args[index]));
+        }
+        index -= args.len();
+        if let Some(vararg) = vararg {
+            if index == 0 {
+                return Some(AnyParameterRef::Variadic(vararg));
+            }
+            index -= 1;
+        }
+        if index < kwonlyargs.len() {
+            return Some(AnyParameterRef::NonVariadic(&kwonlyargs[index]));
+        }
+        if let Some(kwarg) = kwarg.as_deref() {
+            Some(AnyParameterRef::Variadic(kwarg))
+        } else {
+            unreachable!()
+        }
+    }
+
+    fn size_hint(&self) -> (usize, Option<usize>) {
+        let remaining = self.parameters.len() - self.index;
+        (remaining, Some(remaining))
+    }
+}
+
+impl<'a> ExactSizeIterator for ParametersIterator<'a> {}
+impl<'a> FusedIterator for ParametersIterator<'a> {}
+
+impl<'a> IntoIterator for &'a Parameters {
+    type IntoIter = ParametersIterator<'a>;
+    type Item = AnyParameterRef<'a>;
+    fn into_iter(self) -> Self::IntoIter {
+        self.iter()
+    }
+}
+
 /// An alternative type of AST `arg`. This is used for each function argument that might have a default value.
 /// Used by `Arguments` original type.
 ///
diff --git a/crates/ruff_python_parser/src/parser/statement.rs b/crates/ruff_python_parser/src/parser/statement.rs
index 80603ecbc..69d7ec8a5 100644
--- a/crates/ruff_python_parser/src/parser/statement.rs
+++ b/crates/ruff_python_parser/src/parser/statement.rs
@@ -3374,7 +3374,7 @@ impl<'src> Parser<'src> {
         let mut all_arg_names =
             FxHashSet::with_capacity_and_hasher(parameters.len(), BuildHasherDefault::default());
 
-        for parameter in parameters.iter() {
+        for parameter in parameters {
             let range = parameter.name().range();
             let param_name = parameter.name().as_str();
             if !all_arg_names.insert(param_name) {
```

</details>

What do you think?

Reply:
> If we want to be picky, then implementing `ExactSizeIterator` isn't possible because adding the length of two vectors can, in theory, overlap.

I had thought that there was a hard limit of 255 on the number of parameters a Python function was allowed to have. But it seems like that was actually removed many versions ago. So this indeed isn't safe; thanks!

Reply:
We could argue that implementing `ExactSizeIterator` is safe for us (by calling .unwrap() on the `checked_add chain`). it is safe because no program with a max text size of 4GB can have more than u32 parameters. However, that's only true on platforms where `usize` is at least a `u32`.